_posts/2022-01-21-skyhook-bringing-computation-to-storage-with-apache-arrow.md

```diff
@@ -0,0 +1,219 @@
+
+In particular, Skyhook builds on [Ceph][ceph], a distributed storage system that scales to exabytes of data while being reliable and flexible.
+With its Object Class SDK, Ceph enables programmable storage by allowing extensions that define new object types with custom functionality.
+
+## Skyhook Architecture
+
+Let’s look at how Skyhook applies these ideas.
+Overall, the idea is simple: the client should be able to ask Ceph to perform basic operations like decoding files, filtering the data, and selecting columns.
+That way, the work gets done using existing storage cluster resources, which means it’s both adjacent to the data and can scale with the cluster size.
+Also, this reduces the data transferred over the network, and of course reduces the client workload.
+
+On the storage system side, Skyhook uses the Ceph Object Class SDK to define scan operations on data stored in Parquet or Feather format.
+To implement these operations, Skyhook first implements a file system shim in Ceph’s object storage layer, then uses the existing filtering and projection capabilities of the Arrow Datasets library on top of that shim.
+
+Then, Skyhook defines a custom “file format” in the Arrow Datasets layer.
+Queries against such files get translated to direct requests to Ceph using those new operations, bypassing the traditional POSIX file system layer.
+After decoding, filtering, and projection, Ceph directly sends the client Arrow record batches, minimizing CPU overhead for encoding/decoding—another optimization Arrow makes possible.
+The record batches use Arrow’s compression support to further save bandwidth.
+
+<figure>
+  <img src="{{ site.baseurl }}/img/20220121-skyhook-architecture.png"
+       alt="Skyhook Architecture"
+       width="100%" class="img-responsive">
+  <figcaption>
+    Skyhook extends Ceph and Arrow Datasets to push queries down to Ceph, reducing the client workload and network traffic.
+  </figcaption>
+</figure>
+
+Skyhook also optimizes how Parquet files in particular are stored.
+Parquet files consist of a series of row groups, which each contain a chunk of the rows in a file.
+When storing such files, Skyhook either pads or splits them so that each row group is stored as its own Ceph object.
+By striping or splitting the file in this way, we can parallelize scanning at sub-file granularity across the Ceph nodes for further performance improvements.
+
+## Applications
+
+In benchmarks, Skyhook has minimal storage-side CPU overhead and virtually eliminates client-side CPU usage.
+And scaling the storage cluster decreases query latency commensurately.
+For systems like Dask that use the Arrow Datasets API, this means that just by switching to the Skyhook file format, we can speed up dataset scans, reduce the amount of data that needs to be transferred, and free up CPU resources for computations.
+
+<!-- TODO: add the updated figure from Jayjeet -->
```

Comment:
CC @JayjeetAtGithub you mentioned you have an updated version of the CPU usage chart?

Reply:
[cpu_client.pdf](https://github.com/apache/arrow-site/files/7872314/cpu_client.pdf)

[cpu_storage.pdf](https://github.com/apache/arrow-site/files/7872315/cpu_storage_stacked.pdf)

@lidavidm  Here are the updated plots.

Reply:
Thanks! I'll get those integrated. Is there a citation we should put for those?

(Also they look a little less clear-cut than the previous one - is there info on the benchmark scenario? For instance, I assume the client CPU is still being consumed because there's actual work the client is doing?)

Reply:
No, there is no citation for those, those experiments were conducted by us. The client CPU is still getting consumed a little bit, because there is decompression from LZ4 record batches on the client side.

Reply:
Sending back compressed data is actually kind of important in Skyhook, because otherwise we would end up sending back uncompressed Arrow data which is a lot larger than Parquet bytes, and that harms performance adversely.  so, essentially, this is a sort of tradeoff we have to deal with.

Reply:
Ah, thanks. I've tweaked the wording here, if that looks good.

Reply:
I see it. Looks good. Thanks.